libs/testserver/websocket_server.go

```diff
@@ -0,0 +1,66 @@
+package testserver
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/gorilla/websocket"
+)
+
+var upgrader = websocket.Upgrader{
+	ReadBufferSize:  1024,
+	WriteBufferSize: 1024,
+}
+
+type WebsocketServer struct {
+	server *httptest.Server
+	conn   *websocket.Conn
+}
+
+func (s *WebsocketServer) Start() {
+	s.server.Start()
+}
+
+func (s *WebsocketServer) Close() {
+	s.conn.Close()
+	s.server.Close()
+}
+
+func (s *WebsocketServer) Addr() string {
+	return s.server.Listener.Addr().String()
+}
+
+func (s *WebsocketServer) wsHandler(w http.ResponseWriter, r *http.Request) {
+	conn, err := upgrader.Upgrade(w, r, nil)
+	s.conn = conn
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	defer conn.Close()
+
+	for {
+		_, message, err := conn.ReadMessage()
+		if err != nil {
+			fmt.Println(err)
```

Comment:
Can `WebsocketServer` take in a `testing` object and fail if there's an error?

Reply:
Or do a `t.Log` -- the `fmt.Println`'s will disappear in test logs.

Reply:
Good point, updated to use t.Log